CoreAlgorithmPlugins/src/au/gov/asd/tac/constellation/plugins/algorithms/clustering/hierarchical/HierarchicalControllerTopComponent.java

```diff
@@ -719,8 +719,10 @@ public void edit(final GraphWriteMethods graph, final PluginInteraction interact
             final int transactionVisibilityAttribute = VisualConcept.TransactionAttribute.VISIBILITY.ensure(graph);
 
             int nextCluster = 0;
+            // Use the smaller of ints as to avoid going out of bounds.
+            // usually caused when cluster has not run with new elements added.
+            final int vertexCount = Math.min(graph.getVertexCount(), state.getCurrentNumOfClusters());
```

Comment:
The `vertexCount` is being used to retrieve vertex ids, so setting it to the number of clusters feels wrong, and like it could be error prone. What was you reasoning behind this?

Reply:
@cygnus-x-1  So it grabs the group based on the current vertex position it iterates on. My thoughts based on this is that you never want it to grab a group by an iterator that is larger than the total number of groups. This happens when you cluster it once, then add elements to the graph. When you switch off interactive mode it iterates all nodes which were effected in the last cluster so it can revert the changes. Just writing this now it seems like there may be a possibility that it iterates the vertices, but does so under the assumption that the first x elements are the same as the ones which were clustered. I did not notice any issues with testing under normal circumstances. 
Making it the minimum of both makes it avoid checking the groups that does not exist, but still effectively returns the state to normal in the vertices that were effected under "interactive mode" 
Does this sound logical, or do you suggest revisiting it. I didn't want to just re-cluster the graph in case it messed with the behaviour a user would want to target.

Reply:
If I understand correctly, the underlying issue is that a graph could be updated (nodes added, removed, etc) which renders the clustering result invalid. If this is the case I feel like the clustering result should either be removed (forcing the user to re-cluster) or automatically updated (could have an impact on performance). @arcturus2, @sirius-1 what do you think?

Reply:
You do understand correctly. Keen to hear their thoughts. 👍

Reply:
If the cluster is considered invalid then I would prefer we force the user to re-cluster by updating the GUI to look like it wasn't clustered. Perhaps a label somewhere in red which says something like "You need to re-cluster because the graph has changed" would be a nice UX touch.

@cygnus-x-1 given our view of Constellation is to not do "black magic" (i.e. doing things behind the scenes without the user knowing), automatically updating the cluster I think would be going down this path. Would you agree?
Also if automatically updating clusters as the graph changes could be a performance hit then that's a turn off for me personally.

Reply:
While I agree with the "no black magic" principle, in this case the user had asked for the clustering to be done so i wouldn't have an issue with the view keeping that in sync with the graph.